 design/README.MD

```diff
@@ -0,0 +1,456 @@
+can be translated into multiple interoperable programs.
+
+Dafny itself has a `string` type
+that you might assume can map
+to the obvious string type in each translated language.
+
+However a native `string` may be UTF8, UTF16/USC2, or UTF32 
+depending on the language.
+Numbers are similarly complicated.
+Some languages have different sized numbers,
+while others have either a different default or a single size.
+There are also complicated types like streams.
+All these languages have equivalent
+but not congruent types to each other.
+
+In this project Dafny and Smithy are combined
+to generate concrete interfaces in Dafny
+and supported Dafny translation targets.
+These generated interfaces all use the same translated Dafny implementation
+and expose native interfaces and types for each language.
+
+The Smithy shapes are the common denominator
+between all these different types/conversions.
+
+Concretely our traditional greeting program:
+`() => "Hello World!"`
+generate interfaces that can be used like this:
+1. Dafny -> `var hello: seq<char> :- greet();`
+1. Java -> `String hello = greet();`
+1. C# -> `string hello = greet();`
+1. Python -> `hello = greet()`
+1. Rust -> `let hello: String = greet();`
+
+## Definitions
+
+- Runtime or native runtime:
+    A Dafny translation target language,
+    e.g Java or C#.
+    This also may include additional context
+    like language version or other versions/features.
+- Native type or native interface:
+    A type/interface in a runtime,
+    e.g. Java `String`
+- Smithy-runtime:
+    The project used to generate native types
+    for Smithy shapes.
+- Polymorph or polymorph-runtime:
+    The transformation layer between Dafny types
+    and native types,
+    e.g. polymorph-java.
+- Shim:
+    The object that exports the native interface
+    and connects to the internal Dafny implementation.
+- Type converters, ToDafny, ToNative:
+    The dual functions that convert between
+    Dafny and native types.
+    e.g. `D == ToDafny(ToNative(D))`,
+    `N == ToNative(ToDafny(N))`
+
+### Conventions used in this document
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
+in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## Smithy-Dafny modes
+
+### Library
+
+`smithy-dafny` supports `@localService` development.
+This is where the implementation is locally executing code.
+Rather than a remote service like most AWS Service Smithy models.
+
+In this mode `smithy-dafny` will produce Dafny types and interfaces
+for a Smithy model in Dafny.
+This creates an `abstract module` in Dafny you can refine
+with a concrete implementation.
+
+`smithy-dafny` also generates polymorph code
+that will expose native types and interfaces
+for your Smithy model.
+This creates a shim with native interfaces.
+These native interfaces are connected
+to the Dafny interface
```

Comment:
just fyi - this part took me some time to digest, but that was probably due to my unfamiliarity with Dafny. e.g. "we're calling Dafny code from Rust? how?" for me, as a newcomer, it would have been clearer to have a term that clarified it was the "native but non-idiomatic code based on Dafny's typesystem / assertions" (if that's the right way to think of it).